core/src/main/scala/org/apache/spark/api/python/PythonEnvSetup.scala

```diff
@@ -0,0 +1,110 @@
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.SparkContext
+import org.apache.spark.SparkFiles
+import org.apache.spark.internal.config.{PYSPARK_DRIVER_PYTHON, PYSPARK_PYTHON}
+
+
+
+object PythonEnvSetupUtils {
+
+  def getAllSparkNodesReadablePythonEnvRootDir(): Option[String] = {
+    val sparkContext = SparkContext.getActive.get
+    val sparkConf = sparkContext.conf
+
+    // For databricks runtime,
+    // we need to set `allSparkNodesReadableTempDir` to be REPL temp directory
+    val allSparkNodesReadableTempDir = sparkConf.getOption("spark.nfs.rootDir")
+      .orElse {
+        if (sparkContext.isLocal || sparkContext.master.startsWith("local-cluster[")) {
+          // For local / local-cluster mode spark
+          Some(SparkFiles.getRootDirectory())
+        } else None
+      }
+
+    allSparkNodesReadableTempDir.map { path =>
+      val rootEnvDir = new File(path, "spark-udf-python-env-root")
+      rootEnvDir.mkdirs()
+      rootEnvDir.getPath
+    }
+  }
+
+
+  def setupPythonEnvOnSparkDriverIfAvailable(pipDependencies: Seq[String],
+                                             pipConstraints: Seq[String]): Unit = {
+
+    val allSparkNodesReadablePythonEnvRootDirOpt = getAllSparkNodesReadablePythonEnvRootDir()
+
+    if (allSparkNodesReadablePythonEnvRootDirOpt.isDefined) {
+      // If we have python-env root directory that is accessible to all spark nodes,
+      // we can create python environment with provided dependencies in driver side
+      // and spark exectuor can directly use it.
+      val sparkContext = SparkContext.getActive.get
+      val sparkConf = sparkContext.conf
+
+      sparkContext.setLocalProperty(
+        "pythonEnv.allSparkNodesReadableRootEnvDir",
+        allSparkNodesReadablePythonEnvRootDirOpt.get
+      )
+
+      val pythonExec = sparkConf.get(PYSPARK_DRIVER_PYTHON)
+        .orElse(sparkConf.get(PYSPARK_PYTHON))
+        .orElse(sys.env.get("PYSPARK_DRIVER_PYTHON"))
+        .orElse(sys.env.get("PYSPARK_PYTHON"))
+        .getOrElse("python3")
+
+      if (!pipDependencies.isEmpty) {
+        PythonEnvManager.getOrCreatePythonEnvironment(
+          pythonExec,
+          allSparkNodesReadablePythonEnvRootDirOpt.get,
+          pipDependencies,
+          pipConstraints
+        )
+      }
+    }
+  }
+
+  def getPipRequirementsFromChainedPythonFuncs(
+                                                funcs: Seq[ChainedPythonFunctions]
+                                              ): (Seq[String], Seq[String]) = {
+    val headPipDeps = funcs.head.funcs.head.pipDependencies
+    val headPipConstraints = funcs.head.funcs.head.pipConstraints
+
+    for (chainedFunc <- funcs) {
+      for (simpleFunc <- chainedFunc.funcs) {
+        if (simpleFunc.pipDependencies != headPipDeps ||
+          simpleFunc.pipConstraints != headPipConstraints
+        ) {
+          // TODO: For this case, we should split current python runner
+          //  into multiple python runners.
```

Comment:
CC @HyukjinKwon

Reply:
maybe we could first detect whether the two envs can be merged, e.g.:
```
env1:
numpy

env2:
numpy
scipy
```

Reply:
Does this TODO conflict with https://github.com/databricks/runtime/pull/55522 ?

Reply:
Btw, this TODO case is a minor case, usually UDFs in one spark job only requires one kind of python dependency settings.